File: src/java/classify/linkedlist/RemoveNthNodeFromEndOfList.java
package java.classify.linkedlist;

import java.tools.ListNode;

public class RemoveNthNodeFromEndOfList {

    /**
     * 0ms 100.00%
     * 34.1MB 100.00%
     */
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode fast = head;
        ListNode slow = head;

        while (n > 0) {
            fast = fast.next;
            n--;
        }

        if (fast == null) {
            return head.next;
        }

        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }

        slow.next = slow.next.next;

        return head;
    }
}

File: src/python/linked_list/remove_nth_node_from_end_of_list.py
from python.tools.node import ListNode


def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
    """
    52ms 93.76%
    13.1MB 83.1%
    :param self:
    :param head:
    :param n:
    :return:
    """
    if not head:
        return head

    dummy = ListNode(0)
    dummy.next = head
    fast = dummy

    while n:
        fast = fast.next
        n -= 1

    slow = dummy

    while fast and fast.next:
        fast = fast.next
        slow = slow.next

    slow.next = slow.next.next

    return dummy.next
